jsonpatcher/jsonpatcher.go

```diff
@@ -0,0 +1,205 @@
+package jsonpatcher
+
+import (
+	"bytes"
+	"encoding/binary"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"time"
+
+	jsonpatch "github.com/evanphx/json-patch"
+	ds "github.com/ipfs/go-datastore"
+	logging "github.com/ipfs/go-log"
+	es "github.com/textileio/go-eventstore"
+)
+
+type operationType int
+
+const (
+	create operationType = iota
+	save
+	delete
+)
+
+var (
+	log                           = logging.Logger("jsonpatcher")
+	errSavingNonExistentInstance  = errors.New("can't save nonexistent instance")
+	errCantCreateExistingInstance = errors.New("cant't create already existent instance")
+	errUnknownOperation           = errors.New("unknown operation type")
+)
+
+type operation struct {
+	Type      operationType
+	EntityID  es.EntityID
+	JSONPatch []byte
+}
+
+type patcher struct {
+}
+
+var _ es.EventCreator = (*patcher)(nil)
+
+func New() es.EventCreator {
+	return &patcher{}
+}
+
+func (m *patcher) Create(actions []es.Action) ([]es.Event, error) {
+	events := make([]es.Event, len(actions))
+	for i := range actions {
+		var eventPayload []byte
+		var err error
+		switch actions[i].Type {
+		case es.Create:
+			eventPayload, err = createEvent(actions[i].EntityID, actions[i].Current)
+		case es.Save:
+			eventPayload, err = saveEvent(actions[i].EntityID, actions[i].Previous, actions[i].Current)
+		case es.Delete:
+			eventPayload, err = deleteEvent(actions[i].EntityID)
+		default:
+			panic("unkown action type")
+		}
+		if err != nil {
+			return nil, err
+		}
+		events[i] = patchEvent{
+			Timestamp: time.Now(),
+			ID:        actions[i].EntityID,
+			TypeName:  actions[i].EntityType,
+			Patch:     eventPayload,
+		}
+	}
+	return events, nil
+}
+
+func (p *patcher) Reduce(e es.Event, datastore ds.Datastore, baseKey ds.Key) error {
```

Comment:
Here's basically the inverse of `jsonpatcher.Create()`.  `jsonpatcher.Create()` and `jsonpatcher.Reduce()` are the back and forth flow of changes in the model. Could be understood as `encode` and `decode` respectively.

Reply:
If you go with EventCodec, you might want to consider _using_ those names?

Reply:
This is a really nice clear implementation here... really like it. Easy to follow.